File: validation/cases/cubic_spline_natural/cubic_spline_natural.cpp
// cubic_spline_natural.cpp -- Natural cubic spline matching cubic_spline_natural.m
// Usage: ./cubic_spline_natural > cubic_spline_natural_cpp.csv

#include "ctrlpp/trajectory/cubic_spline.h"

#include <cstdio>
#include <vector>

int main()
{
    ctrlpp::cubic_spline<double>::config cfg;
    cfg.times = {0.0, 1.0, 2.5, 4.0, 5.0};
    cfg.positions = {0.0, 1.5, 0.8, 2.0, 1.0};
    cfg.bc = ctrlpp::boundary_condition::natural;

    ctrlpp::cubic_spline<double> spline(cfg);

    constexpr double dt_eval = 0.05;

    std::printf("time,position,velocity,acceleration\n");

    for(double t = cfg.times.front(); t <= cfg.times.back() + dt_eval / 2; t += dt_eval)
    {
        auto pt = spline.evaluate(t);
        std::printf("%.15e,%.15e,%.15e,%.15e\n", t, pt.position(0), pt.velocity(0), pt.acceleration(0));
    }
}

File: validation/cases/cubic_spline_natural/cubic_spline_natural.m
% cubic_spline_natural.m -- Natural cubic spline via Octave built-in spline() + ppder().
%
% Usage: octave --no-gui cubic_spline_natural.m > cubic_spline_natural_octave.csv

% Waypoints
t_knots = [0.0, 1.0, 2.5, 4.0, 5.0];
q_knots = [0.0, 1.5, 0.8, 2.0, 1.0];

% Octave spline() uses not-a-knot BCs by default.
% For natural BCs (zero second derivative at endpoints), use csape from splines pkg.
pkg load splines;
pp = csape(t_knots, q_knots, 'variational');  % 'variational' = natural BCs

% Derivatives via ppder
pp_vel = ppder(pp);
pp_acc = ppder(pp_vel);

% Evaluate at fine grid
dt_eval = 0.05;

fprintf('time,position,velocity,acceleration\n');

t = t_knots(1);
while t <= t_knots(end) + dt_eval/2
    pos = ppval(pp, t);
    vel = ppval(pp_vel, t);
    acc = ppval(pp_acc, t);
    fprintf('%.15e,%.15e,%.15e,%.15e\n', t, pos, vel, acc);
    t = t + dt_eval;
end
